compat/v0/compat.go

```diff
@@ -0,0 +1,359 @@
+	if !ok {
+		return ResponseQueryShards{},
+			fmt.Errorf("unexpected type for Bitcoin pubKey: expected pack.String , got %v",
+				state.State[multichain.Bitcoin].Get("pubKey").Type())
+	}
+
+	shards := make([]CompatShard, 1)
+	shards[0] = CompatShard{
+		DarknodesRootHash: "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
+		Gateways: []Gateway{
+			{
+				Asset:  "BTC",
+				Hosts:  []string{"Ethereum"},
+				Locked: "0",
+				Origin: "Bitcoin",
+				PubKey: bitcoinPubkey.String(),
+			},
+			{
+
+				Asset:  "ZEC",
+				Hosts:  []string{"Ethereum"},
+				Locked: "0",
+				Origin: "Zcash",
+				PubKey: bitcoinPubkey.String(),
+			},
+			{
+				Asset:  "BCH",
+				Hosts:  []string{"Ethereum"},
+				Locked: "0",
+				Origin: "BitcoinCash",
+				PubKey: bitcoinPubkey.String(),
+			},
+		},
+		GatewaysRootHash: "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
+		Primary:          true,
+		PubKey:           bitcoinPubkey.String(),
+	}
+
+	resp := ResponseQueryShards{
+		Shards: shards,
+	}
+	return resp, nil
+}
+
+func TxFromV1Tx(t tx.Tx, hash B32, hasOut bool, bindings txengine.Bindings) (Tx, error) {
+	tx := Tx{}
+
+	phash := t.Input.Get("phash").(pack.Bytes32)
+	tx.Autogen.Set(Arg{
+		Name:  "phash",
+		Type:  "b32",
+		Value: B32(phash),
+	})
+
+	ghash := t.Input.Get("ghash").(pack.Bytes32)
+	tx.Autogen.Set(Arg{
+		Name:  "ghash",
+		Type:  "b32",
+		Value: B32(ghash),
+	})
+
+	nhash := t.Input.Get("nhash").(pack.Bytes32)
+	tx.Autogen.Set(Arg{
+		Name:  "nhash",
+		Type:  "b32",
+		Value: B32(nhash),
+	})
+
+	utxo := ExtBtcCompatUTXO{}
+
+	inamount := t.Input.Get("amount").(pack.U256)
+	utxo.Amount = U256{Int: inamount.Int()}
+	utxo.GHash = B32(ghash)
+
+	gpubkey := t.Input.Get("gpubkey").(pack.Bytes)
+	utxo.ScriptPubKey = B(gpubkey)
+
+	btcTxHash := t.Input.Get("txid").(pack.Bytes)
+	btcTxHash32 := [32]byte{}
+	copy(btcTxHash32[:], btcTxHash)
+	//reverse it
+	utxo.TxHash = B32(btcTxHash32)
+
+	btcTxIndex := t.Input.Get("txindex").(pack.U32)
+	utxo.VOut = U32{Int: big.NewInt(int64(btcTxIndex))}
+
+	tx.Autogen.Set(Arg{
+		Name:  "utxo",
+		Type:  "ext_btcCompatUTXO",
+		Value: utxo,
+	})
+
+	tx.In.Set(Arg{
+		Name:  "utxo",
+		Type:  "ext_btcCompatUTXO",
+		Value: utxo,
+	})
+
+	// can't really re-create this correctly
+	// pray it's OK
+	payload := t.Input.Get("payload").(pack.Bytes)
+	tx.In.Set(Arg{
+		Name: "p",
+		Type: "ext_ethCompatPayload",
+		Value: ExtEthCompatPayload{
+			ABI:   []byte("{}"),
+			Value: B(payload),
+			Fn:    []byte{},
+		},
+	})
+
+	nonce := t.Input.Get("nonce").(pack.Bytes32)
+	tx.In.Set(Arg{
+		Name:  "n",
+		Type:  "b32",
+		Value: B32(nonce),
+	})
+
+	to := t.Input.Get("to").(pack.String)
+	toAddr, err := ExtEthCompatAddressFromHex(to.String())
+	if err != nil {
+		return tx, err
+	}
+
+	tx.In.Set(Arg{
+		Name:  "to",
+		Type:  "ext_ethCompatAddress",
+		Value: toAddr,
+	})
+
+	tokenAddrRaw, err := bindings.TokenAddressFromAsset(multichain.Ethereum, multichain.BTC)
```

Comment:
Looks like we only support compat for Bitcoin transactions for now? Are you planning on having a separate PR for the remaining chains?

Reply:
I was of the impression that we only needed to do compat for BTC as it's the only chain that has third party integrations?

Reply:
@TheAncientGoat Can confirm we need compat for all original chains (i.e. BTC, BCH, and ZEC to/from Ethereum).

Reply:
It shouldn't be hard to support BCH and ZEC too. They're identical structurally, and it is probably better safe than sorry.

Reply:
Ok will look into it in a future PR